File: app/src/main/java/com/lin/read/filter/search/GetChapterContentTask.java
package com.lin.read.filter.search;

import android.app.Activity;
import android.os.AsyncTask;
import android.util.Log;

import com.lin.read.filter.scan.StringUtils;

import java.io.IOException;

/**
 * Created by lisonglin on 2018/3/17.
 */

public class GetChapterContentTask extends AsyncTask<Void, Void, Void> {
    private Activity activity;
    private BookChapterInfo bookChapterInfo;

    public GetChapterContentTask(Activity activity, BookChapterInfo bookChapterInfo, OnTaskListener onTaskListener) {
        this.activity = activity;
        this.bookChapterInfo = bookChapterInfo;
        this.onTaskListener = onTaskListener;
    }

    private OnTaskListener onTaskListener;

    public interface OnTaskListener {
        void onSucc(String chapter, String content);

        void onFailed();
    }

    @Override
    protected Void doInBackground(Void... params) {
        try {
            String content = ResolveChapterUtils.getChapterContent(bookChapterInfo);
            while (!bookChapterInfo.isComplete()) {
                Log.e("Test","未读完，继续读!");
                String temp = ResolveChapterUtils.getChapterContent(bookChapterInfo);
                if (StringUtils.isEmpty(temp)) {
                    break;
                } else {
                    content = content + temp;
                }
            }
            result(true, bookChapterInfo.getChapterName(), content);
        } catch (IOException e) {
            e.printStackTrace();
            result(false, null, null);
        }
        return null;
    }

    private void result(final boolean isSuccessful, final String chapter, final String content) {
        if (onTaskListener == null) {
            return;
        }
        this.activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (isSuccessful) {
                    onTaskListener.onSucc(chapter, content);
                } else {
                    onTaskListener.onFailed();
                }
            }
        });
    }
}

File: app/src/main/java/com/lin/read/filter/search/GetChapterContentTask.kt
package com.lin.read.filter.search

import com.lin.read.retrofit.ReadRetrofitService
import com.lin.read.retrofit.RetrofitInstance
import com.lin.read.utils.*
import okhttp3.ResponseBody
import rx.Observer
import rx.schedulers.Schedulers

class GetChapterContentTask(private val bookChapterInfo: BookChapterInfo, private val onTaskListener: OnTaskListener) {
    fun getChapterContent(chapterLink: String = bookChapterInfo.chapterLink, previousChapter: String = "") {
        RetrofitInstance(chapterLink.baseUrl()).create(ReadRetrofitService::class.java).getResponse(chapterLink)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(object : Observer<ResponseBody> {
                    override fun onError(e: Throwable?) {
                        e?.printStackTrace()
                        onTaskListener.onFailed()
                    }

                    override fun onNext(t: ResponseBody?) {
                        t?.let {
                            val bookChapterContent = ReflectUtil.invokeMethod(ReadBookResolveUtil.Companion, "getChapterContentFrom${bookChapterInfo.webType}", BookChapterContent::class.java, ReadResolveBean(bookChapterInfo, null, it))!!
                            bookChapterContent.takeIf { it.isComplete }?.content?.let {
                                onTaskListener.onSucc(StringKtUtil.removeAdsFromContent(previousChapter + it))
                            } ?: getChapterContent(bookChapterContent.nextLink, previousChapter + bookChapterContent.content)
                        }
                    }

                    override fun onCompleted() {

                    }
                })
    }

    interface OnTaskListener {
        fun onSucc(content: String)
        fun onFailed()
    }
}